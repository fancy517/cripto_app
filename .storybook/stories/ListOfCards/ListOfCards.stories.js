import React from 'react';
import { View } from 'react-native';
import ListOfCards from '../../../src/components/ListOfCards';
import { NavigationContainer } from '@react-navigation/native';

const ListOfCardsMeta = {
    title: 'ListOfCards',
    component: ListOfCards,
    args: [{
        csupply: "",
        id: "1",
        market_cap_usd: "",
        msupply: "",
        name: "2",
        nameid: "",
        percent_change_1h: "",
        percent_change_24h: "",
        percent_change_7d: "",
        price_btc: "",
        price_usd: "",
        rank: 1,
        symbol: "",
        tsupply: "",
        volume24: 1,
        volume24a: 1,
    },
    {
        csupply: "2",
        id: "2",
        market_cap_usd: "2",
        msupply: "2",
        name: "2",
        nameid: "2",
        percent_change_1h: "2",
        percent_change_24h: "2",
        percent_change_7d: "2",
        price_btc: "2",
        price_usd: "2",
        rank: 1,
        symbol: "2",
        tsupply: "2",
        volume24: 2,
        volume24a: 2,
    }],
    decorators: [
        (Story) => (
            <View style={{ justifyContent: 'center', flex: 1, width: '100%' }}>
                <NavigationContainer>
                    <Story />
                </NavigationContainer>
            </View>
        ),
    ],
};

export default ListOfCardsMeta;

export const Basic = {};
