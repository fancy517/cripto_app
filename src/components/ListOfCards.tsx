import { FlatList, StyleSheet, Text, View } from "react-native";
import React, { Component } from "react";
import TokenCard from "./TokenCard";
import Api from "../api";

export class ListOfCards extends Component {
  state = {
    currentPage: 0,
    listOfTokens: [],
    criptoApi: new Api(),
  };

  componentDidMount(): void {
    (async () => {
      const listOfTokens = await this.state.criptoApi.getAllCriptoTickets(
        this.state.currentPage
      );
      this.setState({ ...this.state, listOfTokens });
    })();
  }

  async chargeMoreTokens(): Promise<void> {
    const moreTokens = await this.state.criptoApi.getAllCriptoTickets(
      this.state.currentPage + 1
    );
    this.setState({
      ...this.state,
      listOfTokens: [...this.state.listOfTokens, ...moreTokens],
      currentPage: this.state.currentPage + 1,
    });
  }
  render() {
    return (
      <FlatList
        style={style.container}
        data={this.state.listOfTokens}
        onEndReached={() => this.chargeMoreTokens()}
        onEndReachedThreshold={0.1}
        renderItem={({ item, index }) => <TokenCard token={item} key={index} />}
      />
    );
  }
}

const style = StyleSheet.create({
  container: {
    flex: 1,
  },
});

export default ListOfCards;
