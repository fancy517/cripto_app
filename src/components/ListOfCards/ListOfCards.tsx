import { FlatList, StyleSheet, Text, View } from "react-native";
import React, { Component } from "react";
import TokenCard from "../TokenCard";
import Api from "../../api/Api";
import Token from "../../types/Token";

type listOfCardsProps = {
  queryList: number[];
};

type listOfCardsState = {
  currentPage: number;
  criptoApi: Api;
  queryTimeOut: number;
  listOfTokens: Token[];
};

export class ListOfCards extends Component<listOfCardsProps> {
  constructor(props: listOfCardsProps) {
    super(props);
    this.queryTimeout = this.queryTimeout.bind(this);
  }

  state: listOfCardsState = {
    currentPage: 0,
    queryTimeOut: 0,
    criptoApi: new Api(),
    listOfTokens: [],
  };

  componentDidMount(): void {
    this.initTokens();
  }

  componentDidUpdate(
    prevProps: Readonly<listOfCardsProps>,
    prevState: Readonly<listOfCardsState>,
    snapshot?: any
  ): void {
    if (
      this.props.queryList.length > 0 &&
      this.props.queryList !== prevProps.queryList
    ) {
      this.queryTimeout();
    } else if (
      this.props.queryList.length === 0 &&
      this.props.queryList !== prevProps.queryList
    ) {
      this.initTokens();
      this.setState({ ...this.state, currentPage: 0 });
    }
  }

  async initTokens(): Promise<void> {
    const listOfTokens = await this.state.criptoApi.getAllCriptoTickets(0);
    this.setState({ ...this.state, listOfTokens });
  }

  async chargeMoreTokens(): Promise<void> {
    const moreTokens = await this.state.criptoApi.getAllCriptoTickets(
      this.state.currentPage + 1
    );
    this.setState({
      ...this.state,
      listOfTokens: [...this.state.listOfTokens, ...moreTokens],
      currentPage: this.state.currentPage + 1,
    });
  }

  async makeSearchWithQuery(): Promise<void> {
    const queryTokens = await this.state.criptoApi.getCriptoDetails(
      this.props.queryList.slice(0, 10).join(",")
    );
    this.setState({
      ...this.state,
      listOfTokens: queryTokens,
      currentPage: 0,
    });
  }

  queryTimeout() {
    if (this.state.queryTimeOut !== 0) {
      clearTimeout(this.state.queryTimeOut);
    }
    const tempTimeOut = setTimeout(() => this.makeSearchWithQuery(), 1000);
    this.setState({
      ...this.state,
      queryTimeOut: tempTimeOut,
    });
  }
  render() {
    return (
      <FlatList
        style={style.container}
        data={this.state.listOfTokens}
        onEndReached={
          this.props.queryList.length > 0
            ? () => {}
            : () => this.chargeMoreTokens()
        }
        onEndReachedThreshold={0.1}
        renderItem={({ item, index }) => <TokenCard token={item} key={index} />}
      />
    );
  }
}

const style = StyleSheet.create({
  container: {
    flex: 1,
  },
});

export default ListOfCards;
