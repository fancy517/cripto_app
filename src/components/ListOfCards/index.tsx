import ListOfCards from "./ListOfCards";
export default ListOfCards;
