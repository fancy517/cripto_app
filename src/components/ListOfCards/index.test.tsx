import React from "react";
import renderer from "react-test-renderer";
import ListOfCards from "./index";
import { NavigationContainer } from "@react-navigation/native";
import Token from "../../types/Token";

const mockTokens: Token[] = [
  {
    csupply: "",
    id: "1",
    market_cap_usd: "",
    msupply: "",
    name: "2",
    nameid: "",
    percent_change_1h: "",
    percent_change_24h: "",
    percent_change_7d: "",
    price_btc: "",
    price_usd: "",
    rank: 1,
    symbol: "",
    tsupply: "",
    volume24: 1,
    volume24a: 1,
  },
  {
    csupply: "",
    id: "2",
    market_cap_usd: "",
    msupply: "",
    name: "2",
    nameid: "",
    percent_change_1h: "",
    percent_change_24h: "",
    percent_change_7d: "",
    price_btc: "",
    price_usd: "",
    rank: 2,
    symbol: "",
    tsupply: "",
    volume24: 2,
    volume24a: 2,
  },
];

jest.mock("../../api", () => {
  return jest.fn().mockImplementation(() => {
    return {
      getAllCriptoTickets: async () => mockTokens,
    };
  });
});

describe("ListOfCards Tests", () => {
  it("matches snapshot", () => {
    const jsonElement = renderer
      .create(
        <NavigationContainer>
          <ListOfCards />
        </NavigationContainer>
      )
      .toJSON();
    expect(jsonElement).toMatchSnapshot();
  });
});
