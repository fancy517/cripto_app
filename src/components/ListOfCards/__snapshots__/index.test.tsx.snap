// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ListOfCards Tests matches snapshot 1`] = `
<RCTScrollView
  data={[]}
  getItem={[Function]}
  getItemCount={[Function]}
  keyExtractor={[Function]}
  onContentSizeChange={[Function]}
  onEndReached={[Function]}
  onEndReachedThreshold={0.1}
  onLayout={[Function]}
  onMomentumScrollBegin={[Function]}
  onMomentumScrollEnd={[Function]}
  onScroll={[Function]}
  onScrollBeginDrag={[Function]}
  onScrollEndDrag={[Function]}
  removeClippedSubviews={false}
  renderItem={[Function]}
  scrollEventThrottle={50}
  stickyHeaderIndices={[]}
  style={
    {
      "flex": 1,
    }
  }
  viewabilityConfigCallbackPairs={[]}
>
  <View />
</RCTScrollView>
`;
